# Importing Libraries

In [1]:
import pandas as pd
import json, os
import numpy as np

This file is going to take two CSV's from the Superheroes dataset from Kaggle and clean and transform the data into one file

In [2]:
info = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vS1ZstYLwFgwhZnqDsPjtnlHYhJp_cmW55J8JD5mym0seRsaem3px7QBtuFF0LiI7z1PLCkVKAkdO7J/pub?output=csv'
powers = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vSzdWOBaXOoz52vPmCFV5idNlDBohLY1Lsbc1IfZIZQ7cV_aNB2wYBfhF49uE1TaO1B5MQCGWiNrFfd/pub?output=csv'

dfinfo = pd.read_csv(info)
dfpowers = pd.read_csv(powers)

In [3]:
dfinfo.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [4]:
dfpowers.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [5]:
dfinfo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [6]:
dfpowers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


Powers is the larger table, so we will be joining the tables into info with an inner join. We will join on hero names, which need to be split from the first column in info

# Data cleaning and transformation

## Hero & Publisher

In [7]:
herotest = dfinfo['Hero|Publisher'].str.split('|', expand = True)
herotest

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [8]:
dfinfo[['Hero','Publisher']] = dfinfo['Hero|Publisher'].str.split('|', expand = True)
dfinfo.drop(columns = 'Hero|Publisher', inplace = True)
dfinfo.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

Hero and Publisher are split correctly 

## Gender

In [9]:
dfinfo['Gender'].value_counts()

Male       321
Female     134
Unknown      8
Name: Gender, dtype: int64

Gender seems good here

## Eye color

In [10]:
dfinfo['Eye color'].value_counts()

blue                       182
brown                      102
green                       61
red                         38
Unknown                     17
yellow                      13
black                       12
white                       10
grey                         5
hazel                        5
gold                         3
purple                       3
yellow (without irises)      2
amber                        2
violet                       2
white / red                  1
indigo                       1
silver                       1
green / blue                 1
blue / white                 1
yellow / blue                1
Name: Eye color, dtype: int64

Some thin values here but it mostly looks ok, I dont see any issues with the data

## Race

In [11]:
dfinfo['Race'].value_counts()

Human                156
Unknown              145
Mutant                51
Human / Radiation     11
Android                7
Symbiote               7
Alien                  6
Demon                  6
Cyborg                 6
Kryptonian             5
Atlantean              5
Asgardian              5
Inhuman                4
New God                3
Human-Kree             2
Eternal                2
Demi-God               2
Saiyan                 2
Amazon                 2
Human / Altered        2
Metahuman              2
God / Eternal          2
Human-Vulcan           1
Martian                1
Zombie                 1
Korugaran              1
Talokite               1
Clone                  1
Human-Spartoi          1
Animal                 1
Yautja                 1
Human / Cosmic         1
Tamaranean             1
Mutant / Clone         1
Luphomoid              1
Frost Giant            1
Strontian              1
Czarnian               1
Bolovaxian             1
Human-Vuldarian        1


Same here, lots of unique values but no major issues

## Hair Color

In [12]:
dfinfo['Hair color'].value_counts()

Black               127
Blond                74
Brown                69
No Hair              52
Red                  41
White                22
Unknown              18
Auburn               13
Green                 6
Strawberry Blond      5
Purple                4
Brown / White         4
Grey                  4
black                 3
blond                 3
Silver                3
Blue                  2
Orange                2
Pink                  1
Gold                  1
Red / White           1
Indigo                1
Red / Orange          1
Red / Grey            1
Brownn                1
Yellow                1
Orange / White        1
Magenta               1
Black / Blue          1
Name: Hair color, dtype: int64

No issues here

## Height/Weight
Height and Weight are stored in a dictionary called Measurements. Lets parse those out

In [13]:
# Replacing the single quote with a double quote and using json.loads
measuretest = dfinfo['Measurements'].str.replace("'",'"')
measuretest = json.loads(measuretest[1])
measuretest

{'Height': '191.0 cm', 'Weight': '65.0 kg'}

In [14]:
# Applying to the dataset
dfinfo['Measurements'] = dfinfo['Measurements'].str.replace("'",'"')
dfinfo['Measurements'] = dfinfo['Measurements'].apply(json.loads)
dfinfo['Measurements']

0      {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1       {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2       {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3      {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4      {'Height': '193.0 cm', 'Weight': '122.0 kg'}
                           ...                     
458     {'Height': '183.0 cm', 'Weight': '83.0 kg'}
459     {'Height': '165.0 cm', 'Weight': '52.0 kg'}
460      {'Height': '66.0 cm', 'Weight': '17.0 kg'}
461     {'Height': '170.0 cm', 'Weight': '57.0 kg'}
462     {'Height': '185.0 cm', 'Weight': '81.0 kg'}
Name: Measurements, Length: 463, dtype: object

Measurements is now a proper dictionary. We can split it using pd.Series and concatenate it back to the dataframe

In [15]:
h_w  = dfinfo['Measurements'].apply(pd.Series)
h_w

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

Before concatenating the data back to the dataframe I want to split by space and remove the cm and kg, otherwise I wont be able to turn the new columns into numbers

In [16]:
h_w[['Height','d1']] = h_w['Height'].str.split(' ', expand = True)
h_w[['Weight','d2']] = h_w['Weight'].str.split(' ', expand = True)
h_w.drop(columns = ['d1','d2'], inplace = True)

In [17]:
h_w.head()

Height Weight
0  203.0  441.0
1  191.0   65.0
2  185.0   90.0
3  203.0  441.0
4  193.0  122.0

Now to concatenate it back to the dataframe

In [18]:
dfinfo = pd.concat((dfinfo, h_w), axis = 1)
dfinfo.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height Weight  
0      Marvel Comics  203.0  441.0  
1  Dark Horse Comics  191.0   65.0  
2          DC Comics  185.0   90.0  
3      Marvel Comics  203.0  441.0  
4      Marvel Comics  193.0  122.0

In [19]:
dfinfo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Gender        463 non-null    object
 1   Race          463 non-null    object
 2   Alignment     463 non-null    object
 3   Hair color    463 non-null    object
 4   Eye color     463 non-null    object
 5   Skin color    463 non-null    object
 6   Measurements  463 non-null    object
 7   Hero          463 non-null    object
 8   Publisher     463 non-null    object
 9   Height        463 non-null    object
 10  Weight        463 non-null    object
dtypes: object(11)
memory usage: 39.9+ KB


The columns are still objects, lets change them to float

In [20]:
dfinfo['Height'] = dfinfo['Height'].astype(float)
dfinfo['Weight'] = dfinfo['Weight'].astype(float)
dfinfo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Gender        463 non-null    object 
 1   Race          463 non-null    object 
 2   Alignment     463 non-null    object 
 3   Hair color    463 non-null    object 
 4   Eye color     463 non-null    object 
 5   Skin color    463 non-null    object 
 6   Measurements  463 non-null    object 
 7   Hero          463 non-null    object 
 8   Publisher     463 non-null    object 
 9   Height        463 non-null    float64
 10  Weight        463 non-null    float64
dtypes: float64(2), object(9)
memory usage: 39.9+ KB


In [21]:
# Dropping the old column
dfinfo.drop(columns = 'Measurements', inplace = True)

All set!

## Skin color

In [22]:
dfinfo['Skin color'].value_counts()

Unknown       409
green          15
blue            7
red             7
white           5
grey            4
silver          4
purple          3
gold            2
pink            2
yellow          2
black           1
blue-white      1
orange          1
Name: Skin color, dtype: int64

Largely unknown, but this is fine

## Alignment

In [23]:
dfinfo['Alignment'].value_counts()

good       311
bad        132
neutral     17
Unknown      3
Name: Alignment, dtype: int64

This is good too.

## Powers
Powers are stored in a different table and need to be one hot encoded, then joined with the main table. Lets start with one hot encoding the entries first

In [24]:
type(dfpowers.loc[1,'Powers'])

str

In [25]:
# Splitting the column into individual columns for each power
powertest = dfpowers['Powers'].str.split(',', expand = True)
# Filling the nulls with a string for each new column
powertest.fillna('None', inplace = True)
# Converting to a numpy array
powertest = powertest.values
powertest.astype(str)
powertest

array([['Agility', 'Super Strength', 'Stamina', ..., 'None', 'None',
        'None'],
       ['Accelerated Healing', 'Durability', 'Longevity', ..., 'None',
        'None', 'None'],
       ['Agility', 'Accelerated Healing', 'Cold Resistance', ..., 'None',
        'None', 'None'],
       ...,
       ['Agility', 'Stealth', 'Danger Sense', ..., 'None', 'None',
        'None'],
       ['Cryokinesis', 'Telepathy', 'Magic', ..., 'None', 'None', 'None'],
       ['Super Speed', 'Intangibility', 'Time Travel', ..., 'None',
        'None', 'None']], dtype=object)

In [26]:
# Parsing the array for unique values
powertest = np.unique(powertest)
powertest

array(['Accelerated Healing', 'Adaptation', 'Agility',
       'Animal Attributes', 'Animal Control', 'Animal Oriented Powers',
       'Animation', 'Anti-Gravity', 'Astral Projection', 'Astral Travel',
       'Audio Control', 'Banish', 'Biokinesis', 'Camouflage',
       'Changing Armor', 'Clairvoyance', 'Cloaking', 'Cold Resistance',
       'Cryokinesis', 'Danger Sense', 'Darkforce Manipulation',
       'Death Touch', 'Density Control', 'Dexterity',
       'Dimensional Awareness', 'Dimensional Travel', 'Duplication',
       'Durability', 'Echolocation', 'Elasticity', 'Electrical Transport',
       'Electrokinesis', 'Element Control',
       'Elemental Transmogrification', 'Empathy', 'Energy Absorption',
       'Energy Armor', 'Energy Beams', 'Energy Blasts',
       'Energy Constructs', 'Energy Manipulation', 'Energy Resistance',
       'Enhanced Hearing', 'Enhanced Memory', 'Enhanced Senses',
       'Enhanced Sight', 'Enhanced Smell', 'Enhanced Touch',
       'Fire Control', 'Fire Resis

In [27]:
for col in powertest:
    dfpowers[col] = dfpowers['Powers'].str.contains(col)

dfpowers.head()

C:\Users\Ray\AppData\Local\Temp\ipykernel_2480\1147978363.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dfpowers[col] = dfpowers['Powers'].str.contains(col)
C:\Users\Ray\AppData\Local\Temp\ipykernel_2480\1147978363.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dfpowers[col] = dfpowers['Powers'].str.contains(col)
C:\Users\Ray\AppData\Local\Temp\ipykernel_2480\1147978363.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

   Accelerated Healing  Adaptation  Agility  Animal Attributes  \
0                False       False     True              False   
1                 True       False    False              False   
2                 True       False     True              False   
3                False       False    False              False   
4                 True       False    False              False   

   Animal Control  Animal Oriented Powers  Animation  Anti-Gravity  ...  \
0           False                   False      False         False  ...   
1           False                   False      False         False  ...   
2           False                   False      False         False  ...   
3           False                   False      False         False  ...   
4           False                   False       True         False  ...   

   Vision - Thermal  Vision - X-Ray  Vitakinesis  Wallcrawling  Water Control  \
0             False           False        False         False          False   
1             False           False        False         False          False   
2             False           False        False         False          False   
3             False           False        False         False          False   
4             False           False        False         False          False   

   Weapon-based Powers  Weapons Master  Weather Control  Web Creation  \
0                False           False            False         False   
1                False           False            False         False   
2                False            True            False         False   
3                False           False            False         False   
4                False           False            False         False   

   Wind Control  
0         False  
1         False  
2         False  
3         False  
4         False  

[5 rows x 170 columns]

## Join

In [28]:
joined = pd.merge(dfinfo, dfpowers, left_on = 'Hero', right_on = 'hero_names', how = 'inner')
joined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 180 entries, Gender to Wind Control
dtypes: bool(168), float64(2), object(10)
memory usage: 123.0+ KB


In [29]:
joined.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height  Weight  ... Vision - Thermal  \
0         A-Bomb      Marvel Comics   203.0   441.0  ...            False   
1     Abe Sapien  Dark Horse Comics   191.0    65.0  ...            False   
2       Abin Sur          DC Comics   185.0    90.0  ...            False   
3    Abomination      Marvel Comics   203.0   441.0  ...            False   
4  Absorbing Man      Marvel Comics   193.0   122.0  ...            False   

  Vision - X-Ray  Vitakinesis  Wallcrawling  Water Control  \
0          False        False         False          False   
1          False        False         False          False   
2          False        False         False          False   
3          False        False         False          False   
4          False        False         False          False   

   Weapon-based Powers  Weapons Master  Weather Control  Web Creation  \
0                False           False            False         False   
1                False            True            False         False   
2                False           False            False         False   
3                False           False            False         False   
4                False           False            False         False   

   Wind Control  
0         False  
1         False  
2         False  
3         False  
4         False  

[5 rows x 180 columns]

# Questions

## Compare the average weight of super heroes who have Super Speed to those who do not.

In [30]:
joined.groupby('Super Speed')['Weight'].mean()

Super Speed
False    101.773585
True     129.404040
Name: Weight, dtype: float64

The average weight of super heroes with Super Speed is higher, at 129.4 than those without, at 101.8

## What is the average height of heroes for each publisher?

In [31]:
joined.groupby('Publisher')['Height'].mean().sort_values()

Publisher
George Lucas         159.600000
Shueisha             171.500000
Dark Horse Comics    176.909091
Unknown              178.000000
Team Epic TV         180.750000
Star Trek            181.500000
DC Comics            181.923913
Marvel Comics        191.546128
Image Comics         211.000000
Name: Height, dtype: float64

The average height for each publisher is shown above, in cm. We can see that Image Comics has, on average, the tallest heroes, and George Lucas has the shortest. 